include/seqan3/io/alignment_file/input.hpp

```diff
@@ -913,16 +905,26 @@ class alignment_file_input
     template <std::ranges::ForwardRange ref_sequences_t>
     void set_references(typename traits_type::ref_ids & ref_ids, ref_sequences_t && ref_sequences)
     {
-        assert(std::ranges::size(ref_ids) == std::ranges::size(ref_sequences));
+        assert(std::ranges::distance(ref_ids) == std::ranges::distance(ref_sequences));
 
         header_ptr = std::unique_ptr<header_type>{std::make_unique<header_type>(ref_ids)};
         reference_sequences_ptr = &ref_sequences;
 
         // initialise reference map and ref_dict if ref_ids are non-empty
-        for (size_t idx = 0; idx < ref_ids.size(); ++idx)
+        for (int32_t idx = 0; idx < std::ranges::distance(ref_ids); ++idx)
         {
-            header_ptr->ref_id_info.emplace_back(std::ranges::size(ref_sequences[idx]), "");
-            header_ptr->ref_dict[header_ptr->ref_ids()[idx]] = idx;
+            header_ptr->ref_id_info.emplace_back(std::ranges::distance(ref_sequences[idx]), "");
+
+            if constexpr (std::ranges::ContiguousRange<reference_t<typename traits_type::ref_ids>> &&
+                          std::ranges::SizedRange<reference_t<typename traits_type::ref_ids>>)
```

Comment:
for all the cases you also need to check `seqan3::ForwardingRange` otherwise you might also take this branch for ranges that return temporary objects of contiguousrange that then go out of scope, but you are holding pointers to themm inside your span.